blazar/manager/service.py

```diff
@@ -616,6 +646,58 @@ def _create_reservation(self, values):
         db_api.reservation_update(reservation['id'],
                                   {'resource_id': resource_id})
 
+    def _allocation_candidates(self, lease, reservations):
+        """Returns dict by resource type of reservation candidates."""
+        allocations = {}
+
+        for reservation in reservations:
+            res = reservation.copy()
+            resource_type = reservation['resource_type']
+            res['start_date'] = lease['start_date']
+            res['end_date'] = lease['end_date']
+
+            if resource_type not in self.plugins:
+                raise exceptions.UnsupportedResourceType(
+                    resource_type=resource_type)
+
+            plugin = self.plugins.get(resource_type)
+
+            if not plugin:
+                raise common_ex.BlazarException(
+                    'Invalid plugin names are specified: %s' % resource_type)
+
+            candidate_ids = plugin.allocation_candidates(res)
+
+            allocations[resource_type] = [
+                plugin.get(cid) for cid in candidate_ids]
+
+        return allocations
+
+    def _existing_allocations(self, reservations):
+        allocations = {}
+
+        for reservation in reservations:
```

Comment:
A lot of this function is boilerplate from the above function, but that's not a big deal.